### PR TITLE
Replace boost::optional by std_cxx17::optional.

### LINKED ISSUE #8590: Introduce std_cxx17::optional
C++17 includes `std::optional`. We currently use `boost::optional` but what we should probably do for these locations is to introduce a namespace `std_cxx17` that contains either `std::optional` if the compiler has it, or uses `boost::optional` as an alias if it does not.

### PR DESCRIPTION
**Pull request by masterleinad**
Fixes #8590.

### CODE REVIEW

**Comment by masterleinad** [comment]
I replaced the `DEAL_II_WITH_CXX17` check by `__cpp_lib_optional`. That should be more reliable.

File: include/deal.II/base/std_cxx17/optional.h [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+#ifndef dealii_cxx17_optional_h
+#define dealii_cxx17_optional_h
+
+#include <deal.II/base/config.h>
+
+#ifdef __cpp_lib_optional
+#  include <optional>
+#else
+#  include <boost/optional.hpp>
+#endif // DEAL_II_WITH_CXX17
```
Comment (bangerth) [inline include/deal.II/base/std_cxx17/optional.h]: might want to replace it here as well

File: include/deal.II/base/std_cxx17/optional.h [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+DEAL_II_NAMESPACE_OPEN
+namespace std_cxx17
+{
+#ifdef __cpp_lib_optional
+  using std::optional;
+#else
+  using boost::optional;
+#endif // DEAL_II_WITH_CXX17
```
Comment (bangerth) [inline include/deal.II/base/std_cxx17/optional.h]: and here

**Comment by masterleinad** [comment]
> Is `__cpp_lib_optional` one of the C++17 feature macros?

Yes, https://en.cppreference.com/w/User:D41D8CD98F/feature_testing_macros#C.2B.2B17

**Pull request merged**